### ISSUE TITLE
DMTCP seg faults using python

### ISSUE THREAD

**Issue by mamelara** [dmtcp/dmtcp#708]
Hello! Wondering if this is the right place to post issues or on sourceforge. We are trying to get dmtcp installed on our cluster here at NERSC and I am running into an odd issue. 

## Issue
dmtcp seg faults whenever I try to run a command
```console
$ dmtcp_launch spades.py --test
Segmentation fault
```

## Expected 
dmtcp to launch the spades.py --test which results in the following output
```console
$ spades.py --test
Command line: /usr/common/jgi/linux/opensuse42/x86_64/spades/3.12.0/gcc/7.1.0/v45qtin/bin/spades.py     --test

System information:
  SPAdes version: 3.12.0
  Python version: 2.7.12
  OS: Linux-2.6.32-696.18.7.el6.nersc.x86_64-x86_64-with-SuSE-42.2-x86_64

Output dir: /global/u2/m/mamelara/spades_test
Mode: read error correction and assembling
Debug mode is turned OFF

Dataset parameters:
  Multi-cell mode (you should set '--sc' flag if input data was obtained with MDA (single-cell) technology or --meta flag if processing metagenomic dataset)
  Reads:
    Library number: 1, library type: paired-end
      orientation: fr
      left reads: ['/global/common/genepool_sl72/jgi/linux/opensuse42/x86_64/spades/3.12.0/gcc/7.1.0
```

I've noticed that this actually happens even with simple hello world scripts 

```console
$ dmtcp_launch ./hello_world.py
Segmentation fault
```

## Debug info
```console
$ gdb --args dmtcp_launch /usr/common/jgi/linux/opensuse42/x86_64/spades/3.12.0/gcc/7.1.0/v45qtin/bin/spades.py --test
Starting program: /global/common/genepool_sl72/jgi/linux/opensuse42/x86_64/dmtcp/2.5.2/gcc/7.1.0/fcod4ja/bin/dmtcp_launch /usr/common/jgi/linux/opensuse42/x86_64/spades/3.12.0/gcc/7.1.0/v45qtin/bin/spades.py --test
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
warning: File "/global/common/genepool_sl72/usg/languages/gcc/7.1.0/lib64/libstdc++.so.6.0.23-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
To enable execution of this file add
        add-auto-load-safe-path /global/common/genepool_sl72/usg/languages/gcc/7.1.0/lib64/libstdc++.so.6.0.23-gdb.py
line to your configuration file "/global/homes/m/mamelara/.gdbinit".
To completely disable this security protection add
        set auto-load safe-path /
line to your configuration file "/global/homes/m/mamelara/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
        info "(gdb)Auto-loading safe path"
process 14472 is executing new program: /os/usr/bin/env
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
warning: File "/global/common/genepool_sl72/usg/languages/gcc/7.1.0/lib64/libstdc++.so.6.0.23-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
[New Thread 0x2aaaae2d8700 (LWP 14594)]

Thread 1 "spades.py" received signal SIGSEGV, Segmentation fault
hash_first (use_gnu_hash=0, hash_table=0xffffffffff700124, name=0x2aaaab7cfdcf "__vdso_clock_gettime") at dmtcp_dlsym.cpp:145
145     dmtcp_dlsym.cpp: No such file or directory.
```

Not sure why that error is happening, dmtcp_dlsym.cpp would have been linked into the executable right? Not sure why it's not finding it. 

## System information
OS: OpenSuSE42
Compiler: gcc/7.1.0
Python: 2.7-anaconda-4.4  

Thanks for the help. Wonder if this is related to #588?

**Comment by JainTwinkle** [comment]
Which version of DMTCP are you using?
I am using DMTCP 3.0 at my local PC and not seeing the same error. I do have an account on cori. However, Cori is down at this moment. I can try to reproduce once Cori is up.

**Comment by mamelara** [comment]
2.5.2 is the version for the latest release. Is the current git repository v3.0? This is on our Denovo cluster which is different than Cori. I have also considered that this could be a system config issue as well but I'm not 100% certain. 

I'll try out 3.0.

**Comment by mamelara** [comment]
Update: Looks like version 3.0 also gives the same seg fault

**Comment by gc00** [comment]
Rohan,
    I vaguely remember you telling me that there was a single, particular
version of Python (or other package?) that created a problem for DMTCP,
but that all other version (before or after) worked fine.  Do I remember
this correctly?

Hi Mario,
    While we're waiting on Rohan's answer, I have a thought on how to
proceed if my memory is correct.  It's actually not hard to download
and compile a new Python (i.e., a newer version).  You could do a quick
test of helloWorld.py based on a newer Python -- just to see if it's
something strange with this one version of Python.
    Normally, we don't see any problems from our users in checkpointing
Python.

Best,
- Gene

----- Original Message -----

**Comment by gc00** [comment]
Rohan (@rohgarg),
    Does this strace output, below, remind you of anything?
Twinkle,
    Were you able to test Python on Cori/NERSC yet, to see if
you can reproduce the issue?

Best,
- Gene

----- Original Message -----

**Comment by JainTwinkle** [comment]
@gc00  I tried to reproduce the same error on Cori (both login and compute node). But, I don't see any error on Cori. According to @mamelara, Cori is different than their system (Denovo). So, I am not sure whether running on Cori would help much.
--Twinkle

**Comment by mamelara** [comment]
Hi yes,

Working on Cori is our workaround but it's still a mystery to me as to what would be the main problem between Denovo and Cori. I could maybe grant access to Denovo if you already have a NERSC account just to test it out. 

Also upon Cori returning one of our users decided to run dmtcp_launch with the example shown above and noticed a significant difference in the time it takes to run and complete the command: 

```console
#set up dmtcp env
  module load dmtcp
#set path to install of spades with shared libraries
 PATH=/global/cscratch1/sd/aclum/software/cray-cnl6-haswell/gcc-7.3.0/spades-3.12.0-3gqftu3a7prbctcyjd57xqyigave6na2/bin:$PATH
#test w/o checkpointing, completes in less than 1 min
 /usr/bin/time spades.py --test

# test with checkpointing, hangs for 15 min before I killed it.
/usr/bin/time dmtcp_launch spades.py --test
``` 

I can open a separate issue about this

**Comment by JainTwinkle** [comment]
@mamelara If you could grant access to Denovo then it would be easy for me to debug the issues. My username is `tjain`.
Thanks,
--Twinkle

**Comment by mamelara** [comment]
@JainTwinkle are you able to access Denovo? No rush of course. Just want to make sure everything works for you.

**Comment by JainTwinkle** [comment]
@mamelara Yes, I was able to access denovo. I tried as you said. 
`
$ module load dmtcp/2.5.2 
$ dmtcp_launch --version
dmtcp_launch (DMTCP) 2.5.2
License LGPLv3+: GNU LGPL version 3 or later
    <http://gnu.org/licenses/lgpl.html>.
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; see COPYING file for details.
$ dmtcp_launch python 
Python 2.7.15 |Anaconda custom (64-bit)| (default, May  1 2018, 23:32:55) 
[GCC 7.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> a = 10
>>> a
10
>>> 
$ dmtcp_launch ./dmtcp_master/dmtcp/test/hello_world.py 
hello world
`

However, I **am able to reproduce** the same error as you are getting with `spades` module on Denovo. And, yes the issue is same as #588 

`#0  0x00007ffff70195d8 in hash_first (name=0x7ffff705859c "__vdso_clock_gettime", hash_table=0xffffffffff700124, use_gnu_hash=0) at dmtcp_dlsym.cpp:166
#1  0x00007ffff7019e10 in dlsym_default_internal_library_handler (handle=0x7ffff7ffe6e0, symbol=0x7ffff705859c "__vdso_clock_gettime", version=0x0, 
    tags_p=0x7fffffff9cd0, default_symbol_index_p=0x7fffffff9ccc) at dmtcp_dlsym.cpp:368
#2  0x00007ffff701a472 in dlsym_default_internal_flag_handler (handle=0x0, libname=0x7ffff70585b1 "linux-vdso", 
    symbol=0x7ffff705859c "__vdso_clock_gettime", version=0x0, addr=0x7ffff70302ae <dmtcp::ProcessInfo::serialize(jalib::JBinarySerializer&)+614>, 
    tags_p=0x7fffffff9cd0, default_symbol_index_p=0x7fffffff9ccc) at dmtcp_dlsym.cpp:478
#3  0x00007ffff701a730 in dmtcp_dlsym_lib_fnc_offset (libname=0x7ffff70585b1 "linux-vdso", symbol=0x7ffff705859c "__vdso_clock_gettime")
    at dmtcp_dlsym.cpp:622
#4  0x00007ffff70302ae in dmtcp::ProcessInfo::serialize (this=0x7ffff7fe3408, o=...) at processinfo.cpp:761
#5  0x00007ffff702c32a in dmtcp::processInfo_EventHook (event=DMTCP_EVENT_PRE_EXEC, data=0x7fffffffaad8) at processinfo.cpp:87
#6  0x00007ffff701ad74 in dmtcp::PluginInfo::eventHook (this=0x7ffff7febc08, event=DMTCP_EVENT_PRE_EXEC, data=0x7fffffffaad8) at plugininfo.cpp:90
#7  0x00007ffff701d13d in dmtcp::PluginManager::eventHook (event=DMTCP_EVENT_PRE_EXEC, data=0x7fffffffaad8) at pluginmanager.cpp:233
#8  0x00007ffff70014b6 in dmtcpPrepareForExec (path=0x7fffffffc10e "python", argv=0x7fffffffbab0, filename=0x7fffffffb758, newArgv=0x7fffffffb750)
    at execwrappers.cpp:408
#9  0x00007ffff70028d2 in execvp (filename=0x7fffffffc10e "python", argv=0x7fffffffbab0) at execwrappers.cpp:643
#10 0x00007ffff77a757f in execvp (filename=0x7fffffffc10e "python", argv=0x7fffffffbab0) at ipc/ssh/ssh.cpp:580
#11 0x0000000000401a4e in ?? ()
#12 0x00007ffff69f66e5 in __libc_start_main () from /lib64/libc.so.6
#13 0x0000000000401ae9 in ?? () `

I will let you know if I find more information.
